### PR TITLE
Basic instrumentation for HTTParty Gem.

### PR DESCRIPTION
**Pull request by pkalmakis**
### HTTParty

The `httparty` integration adds instrumentation for sending HTTP requests using HTTParty.

To setup your integration, use the `Datadog.configure` method like so:

```ruby
require 'httparty'
require 'ddtrace'

Datadog.configure do |c|
  c.use :httparty, options
end
```

Where `options` is an optional `Hash` that accepts the following parameters:

| Key | Description | Default |
| --- | ----------- | ------- |
| `analytics_enabled` | Enable analytics for spans produced by this integration. `true` for on, `nil` to defer to global setting, `false` for off. | `false` |
| `distributed_tracing` | Enables [distributed tracing](#distributed-tracing) | `true` |
| `service_name` | Service name for `httparty` instrumentation. | `'httparty'` |
| `tracer` | `Datadog::Tracer` used to perform instrumentation. Usually you don't need to set this. | `Datadog.tracer` |

In situations where you have multiple clients calling multiple distinct services, you may configure the clients like so:

```ruby
class FooClient
  include HTTParty
  base_uri 'https://example.com'
  dd_options service_name: 'foo_service'

  def foo
    self.class.get('/foo')
  end
end
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -57,3 +57,7 @@ gemfiles/*
 
 # bundle config
 gemfiles/.bundle
+
+# RubyMine/IntelliJ
+.idea
+*.iml
```
Comment (marcotc) [inline .gitignore]: I'd suggest you add these IDE-specific ignore rules under your `~/.gitignore` instead, and configure git to use it like so: `git config --global core.excludesfile '~/.gitignore'`.

File: docs/GettingStarted.md [hunk 2]
```diff
@@ -340,12 +341,13 @@ For a list of available integrations, and their configuration options, please re
 | Grape                    | `grape`                    | `>= 1.0`                 | *[Link](#grape)*                    | *[Link](https://github.com/ruby-grape/grape)*                                  |
 | GraphQL                  | `graphql`                  | `>= 1.7.9`               | *[Link](#graphql)*                  | *[Link](https://github.com/rmosolgo/graphql-ruby)*                             |
 | gRPC                     | `grpc`                     | `>= 1.10`                | *[Link](#grpc)*                     | *[Link](https://github.com/grpc/grpc/tree/master/src/rubyc)*                   |
+| HTTParty                 | `httparty`                 | `>= 0.17`                | *[Link](#httparty)*                 | *[Link](https://github.com/jnunemaker/httparty)*                               |
```
Comment (marcotc) [inline docs/GettingStarted.md]: It's very likely not all our users are up-to-date with their version of `httparty`.

Could you please verify what's the minimum version of `httparty` we can support with the implementation this PR?

Just to be clear, I don't expect you to make changes to accommodate all old versions of `httparty`, but it's very likely that we support older version without changing the implementation, we just have to double check that the code on older versions of `httparty` is compatible.

File: lib/ddtrace/contrib/httparty/integration.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        def self.version
+          Gem.loaded_specs['httparty'] && Gem.loaded_specs['httparty'].version
+        end
+
+        def self.present?
+          super && defined?(::HTTParty)
+        end
+
+        def self.compatible?
+          super && Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
```
Comment (marcotc) [inline lib/ddtrace/contrib/httparty/integration.rb]: I don't think we need to explicitly check the Ruby version here.

This check is only necessary if the user is able to install the gem (`httparty`) with a specific version of Ruby that would impact the integration.

In the case of `httparty`, I don't see any impediments that would hinder tracing, as long as the gem is successfully installed.

File: lib/ddtrace/contrib/httparty/patcher.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        module_function
+
+        def patched?
+          done?(:httparty)
+        end
+
+        def patch
+          do_once(:httparty) do
+            require 'ddtrace/contrib/httparty/module_patch'
+            require 'ddtrace/contrib/httparty/request_patch'
```
Comment (marcotc) [inline lib/ddtrace/contrib/httparty/patcher.rb]: I quite like these `require` statements being delayed up until this point! 💯 

It's definitely something we could do in other parts of the code.

File: lib/ddtrace/contrib/httparty/request_patch.rb [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+            method = http_method::METHOD
+            datadog_trace_request(method, uri) do |span|
+              if distributed_tracing?
+                headers = options[:headers].respond_to?(:to_hash) ? options[:headers].to_hash : {}
+                Datadog::HTTPPropagator.inject!(span.context, headers)
+                options[:headers] = headers
+              end
+
+              super(&block)
```
Comment (marcotc) [inline lib/ddtrace/contrib/httparty/request_patch.rb]: We should make sure any errors that occur in our tracing logic do not affect the application's ability to perform its normal functions.

We might want to have a `rescue` clause for all our code, and make sure that `super` is always called.

File: docs/GettingStarted.md [hunk 6]
```diff
@@ -880,6 +882,44 @@ alternate_client = Demo::Echo::Service.rpc_stub_class.new(
...
+| `service_name` | Service name for `httparty` instrumentation. | `'httparty'` |
+| `tracer` | `Datadog::Tracer` used to perform instrumentation. Usually you don't need to set this. | `Datadog.tracer` |
+
+In situations where you have multiple clients calling multiple distinct services, you may configure the clients like so:
+
+```ruby
+class FooClient
+  include HTTParty
+  base_uri 'https://example.com'
+  dd_options service_name: 'foo_service'
```
Comment (marcotc) [inline docs/GettingStarted.md]: This is really cool.
I'm not 100% sure on the naming here (`dd_options`).
I think a third opinion would be helpful regarding such naming.

File: lib/ddtrace/contrib/httparty/patcher.rb (hunk 4 above)
Comment (pkalmakis) [reply lib/ddtrace/contrib/httparty/patcher.rb]: I think I copied this from another existing integration :)

File: docs/GettingStarted.md (hunk 2 above)
Comment (pkalmakis) [reply docs/GettingStarted.md]: I would say this was one area of general confusion that our contribution guidelines could offer more guidance on.

Comment (pkalmakis) [reply docs/GettingStarted.md]: Looks like we can go as far back as `0.14.0` with this implementation.

File: lib/ddtrace/contrib/httparty/configuration/settings.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      module Configuration
+        # Custom settings for the HTTParty integration
+        class Settings < Contrib::Configuration::Settings
+          option  :analytics_enabled,
+                  default: -> { env_to_bool(Ext::ENV_ANALYTICS_ENABLED, false) },
+                  lazy: true
+
+          option  :analytics_sample_rate,
+                  default: -> { env_to_float(Ext::ENV_ANALYTICS_SAMPLE_RATE, 1.0) },
+                  lazy: true
```
Comment (marcotc) [inline lib/ddtrace/contrib/httparty/configuration/settings.rb]: We've just introduced a new configuration DSL syntax for `option`s, would you mind refactoring these to look similar to this example? https://github.com/DataDog/dd-trace-rb/blob/18300bd2c645c213c22ade2fd66b5bc6151be450/lib/ddtrace/contrib/racecar/configuration/settings.rb#L10-L20

File: spec/ddtrace/contrib/httparty/example_client_spec.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    class FooClient
+      include HTTParty
+      base_uri 'https://example.com'
+      dd_options service_name: 'foo_service'
+
+      def foo
+        self.class.get('/foo')
+      end
+    end
+    FooClient
```
Comment (marcotc) [inline spec/ddtrace/contrib/httparty/example_client_spec.rb]: We try to ensure we don't add new constants to the global name.
In this regard, I suggest creating a new class object inside this `let` block and returning it.

```ruby
let(:klass) do
  Class.new do
    include HTTParty
    base_uri 'https://example.com'
    # ...
  end
end
```

Comment (pkalmakis) [reply spec/ddtrace/contrib/httparty/example_client_spec.rb]: I caught this one too :). There shouldn't be any warnings now.

File: lib/ddtrace/contrib/httparty/module_patch.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+require 'ddtrace/propagation/http_propagator'
+require 'ddtrace/contrib/rest_client/ext'
+
+module Datadog
+  module Contrib
+    module HTTParty
+      # HTTParty ModulePatch
+      module ModulePatch
+        def self.included(base)
+          base.send(:prepend, ClassMethods)
```
Comment (delner) [inline lib/ddtrace/contrib/httparty/module_patch.rb]: Curious; why `prepend` over `include` here?

File: lib/ddtrace/contrib/httparty/module_patch.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/propagation/http_propagator'
+require 'ddtrace/contrib/rest_client/ext'
+
+module Datadog
+  module Contrib
+    module HTTParty
+      # HTTParty ModulePatch
+      module ModulePatch
```
Comment (delner) [inline lib/ddtrace/contrib/httparty/module_patch.rb]: I would suggest a name change to communicate intent over implementation:

```suggestion
      module Helpers
```

Might make more sense to nest this under `Instrumentation` as well e.g.:

 - `Datadog::Contrib::HTTParty::Instrumentation::Helpers`
 - `Datadog::Contrib::HTTParty::Instrumentation::Request`

Similarly the files would have to be renamed and nested to reflect the namespace change too; 
 - `contrib/httparty/instrumentation/helpers.rb`
 - `contrib/httparty/instrumentation/request.rb`

The names are a bit longer, but I think something like that might better communicate purpose. Open to suggestions for other names too!

File: lib/ddtrace/contrib/httparty/request_patch.rb [hunk 11]
```diff
@@ -0,0 +1,137 @@
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/propagation/http_propagator'
+require 'ddtrace/contrib/rest_client/ext'
+
+module Datadog
+  module Contrib
+    module HTTParty
+      # HTTParty RequestPatch
+      module RequestPatch
```
Comment (delner) [inline lib/ddtrace/contrib/httparty/request_patch.rb]: Also see my other suggestion above, which would have this change to something like `Datadog::Contrib::HTTParty::Instrumentation::Request`.

File: lib/ddtrace/contrib/httparty/request_patch.rb [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          def uri
+            # uri is a method not a property in underlying class
+            # cache it so that it is only called once per call to 'perform', and then delete it
+            @dd_uri ||= super
+          end
+
+          def perform(&block)
+            return super unless tracer.enabled
```
Comment (delner) [inline lib/ddtrace/contrib/httparty/request_patch.rb]: We do this a lot in other, older instrumentation, so this is fine, no change required. But reminded me that (although out of scope for this PR) we should have `Tracer#trace` check this instead.

To make that possible, we'd need no-op spans given the current implementation, something we've thought about doing a few times. @marcotc, if we implemented that `tracer.trace { |span| span.instrument { super } }` behavior we discussed, we could do this without no-ops, making lines like this obsolete (you'd just check if enabled and invoke the block if disabled instead of tracing.)

File: lib/ddtrace/contrib/httparty/request_patch.rb [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            end
+          end
+
+          def datadog_finish_trace(span, response)
+            return if span.nil? || !response.is_a?(::HTTParty::Response)
+
+            span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response.code)
+          end
+
+          def datadog_trace_request(method, uri)
```
Comment (delner) [inline lib/ddtrace/contrib/httparty/request_patch.rb]: This method seems very duplicative of `Tracer#trace`: is there a simpler way of doing this? Haven't yet wrapped my head around the instrumentation strategy. If you could offer some brief explanation of how HTTParty performs requests and our instrumentation strategy here, that would be helpful.

File: spec/ddtrace/contrib/httparty/example_client_spec.rb [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+RSpec.describe 'ExampleClient' do
+  let(:tracer) { get_test_tracer }
+  let(:configuration_options) { { tracer: tracer } }
+
+  before do
+    Datadog.configure do |c|
+      c.use :httparty, configuration_options
+    end
+
+    WebMock.disable_net_connect!
```
Comment (delner) [inline spec/ddtrace/contrib/httparty/example_client_spec.rb]: I'm all for preventing outgoing HTTP requests, but we've had a number of problems with WebMock. Is there anyway to use RSpec stubs/mocks to do this in a way that's not overly burdensome or brittle?

File: lib/ddtrace/contrib/httparty/configuration/settings.rb (hunk 7 above)
Comment (pkalmakis) [reply lib/ddtrace/contrib/httparty/configuration/settings.rb]: I'm running into some trouble with this change. When running my tests, `o` is always `nil`.

File: lib/ddtrace/contrib/httparty/module_patch.rb (hunk 9 above)
Comment (pkalmakis) [reply lib/ddtrace/contrib/httparty/module_patch.rb]: Ah. I was playing around with monkey-patching some methods on the `HTTParty` before deciding against it. Changing this to `:include`.

File: lib/ddtrace/contrib/httparty/configuration/settings.rb (hunk 7 above)
Comment (marcotc) [reply lib/ddtrace/contrib/httparty/configuration/settings.rb]: It seems like you need to rebase your branch with the latest master :)

**Comment by pkalmakis** [comment]
In broad strokes, `HTTParty` is a gem that provides a wrapper around `Net::HTTP` with a more developer-friendly API. Typing this out, I'm actually wondering if the `HTTParty` integration should also just use existing `Net::HTTP` instrumentation under the hood. I also want to test what happens if both are enabled at the same time and make sure we don't generate duplicate spans.

**Comment by delner** [comment]
> I also want to test what happens if both are enabled at the same time and make sure we don't generate duplicate spans.

@pkalmakis If the HTTParty span acts as a parent to the Net::HTTP span, I think that would probably be okay. Can you share what a trace with both of those look like?

File: lib/ddtrace/contrib/httparty/request_patch.rb (hunk 5 above)
Comment (pkalmakis) [reply lib/ddtrace/contrib/httparty/request_patch.rb]: Much of the logic within `tracer.trace` does this.

File: lib/ddtrace/contrib/httparty/request_patch.rb (hunk 13 above)
Comment (pkalmakis) [reply lib/ddtrace/contrib/httparty/request_patch.rb]: updated

**Comment by pkalmakis** [comment]
image
This is a trace containing both HTTParty and Net::HTTP spans. Looks fine to me.

File: spec/ddtrace/contrib/httparty/example_client_spec.rb (hunk 14 above)
Comment (pkalmakis) [reply spec/ddtrace/contrib/httparty/example_client_spec.rb]: This has been very stable in my own testing, and to do something with mocks would require a lot of digging through both the HTTParty gem, and Net::HTTP, which at the end of the day could lead to more brittle tests if implementations change in the future.

I made a few attempts at doing some simple mocking, and one tough issue I ran into was that rspec does not allow mocking of methods that have been prepended.

File: docs/GettingStarted.md (hunk 6 above)
Comment (marcotc) [reply docs/GettingStarted.md]: I think `ddtrace_options` would help us avoid possible conflicts in the future, specially with Datadog's own `datadog/statsd` gem.

One other option is to have it as only `ddtrace`, like `ddtrace service_name: 'foo_service'`, but I think this approach doesn't communicate the intention of the method enough.

Having a configuration block, instead of a hash is also in line with our other configuration patterns `ddtrace_options do |o| o.service_name = 'foo_service' end`, but it seems like it would make the user's code less readable.

File: docs/GettingStarted.md [hunk 15]
```diff
@@ -324,11 +325,11 @@ For a list of available integrations, and their configuration options, please re
 
 | Name                     | Key                        | Versions Supported       | How to configure                    | Gem source                                                                     |
 | ------------------------ | -------------------------- | ------------------------ | ----------------------------------- | ------------------------------------------------------------------------------ |
-| Action View              | `action_view`              | `>= 3.2`          | *[Link](#action-view)*              | *[Link](https://github.com/rails/rails/tree/master/actionview)*                |
+| Action View              | `action_view`              | `>= 3.2`                 | *[Link](#action-view)*              | *[Link](https://github.com/rails/rails/tree/master/actionview)*                |
```
Comment (marcotc) [inline docs/GettingStarted.md]: nit: I think it's best to avoid touching these lines that we didn't change in this PR, to ensure a cleaner history trace of this table.

File: docs/GettingStarted.md (hunk 6 above)
Comment (pkalmakis) [reply docs/GettingStarted.md]: I'll update the name to `ddtrace_options`.

**Comment by horaciob** [comment]
Hi there, there is any date or blocking for this PR? there is any way to help with this?

**Comment by marcotc** [comment]
Looking at the source code of HTTParty, it seems like it [always uses the native `Net::HTTP`](https://github.com/jnunemaker/httparty/blob/f35528e240c4ff47df2b7bf4858845b4cb4b423d/lib/httparty/connection_adapter.rb#L99-L113).

Given this, HTTParty can be indirectly instrumented today by enabling the `Net::HTTP` instrumentation (`c.use :http`).

I don't see any information captured by HTTParty that would not also be visible in the `Net::HTTP` layer.

I'm leaning towards not having this integration in `ddtrace` as its own integration.

I do think that we should document how HTTParty users can get APM visibility through `Net::HTTP`, to ensure this is clearly discoverable.

**Comment by ivoanjo** [comment]
As @marcotc commented in https://github.com/DataDog/dd-trace-rb/pull/837#issuecomment-827001276, `Net::HTTP` should cover for httparty usage.

I'm closing this PR for now, please do re-open (or file an issue) if there's any problems with using `Net::HTTP` :)
